File: apps/windows/TeslaSync.App/shared-surfaces/HelpTooltip/HelpTooltip.Model.cs
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.SharedSurfaces.HelpTooltipSurface;

/// <summary>
/// Canonical metadata for the <c>HelpTooltip</c> shared surface — the native mirror of the module-level
/// constants in <c>web/src/components/ui/HelpTooltip.tsx</c>. The web component is a compact "?" affordance that
/// reveals an explanatory tooltip on hover / focus / tap with an optional "Learn more" link, so this carries the
/// diagnostics slug, the two fixed i18n keys it owns (the trigger's default accessible name
/// <c>help.tooltip.iconLabel</c> and the learn-more link's default label <c>common.learnMore</c>) with their
/// verbatim English fallbacks, the Segoe Fluent glyphs that stand in for the web Lucide icons (the
/// <c>HelpCircle</c> trigger and the <c>ExternalLink</c> learn-more affordance), the ARIA role the revealed body
/// declares (web <c>role="tooltip"</c>), and the trigger glyph sizes for the three web size tiers. UI-free so the
/// mapping is asserted in tests without a XAML runtime.
/// </summary>
public static class HelpTooltipRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event (P1/S11 contract).</summary>
    public const string Slug = "HelpTooltip";

    /// <summary>i18n key for the trigger's default accessible name (web <c>t('help.tooltip.iconLabel')</c>).</summary>
    public const string IconLabelKey = "translation.help.tooltip.iconLabel";

    /// <summary>English fallback for <see cref="IconLabelKey"/> (web second arg, verbatim).</summary>
    public const string IconLabelFallback = "More info";

    /// <summary>i18n key for the learn-more link's default label (web <c>t('common.learnMore')</c>).</summary>
    public const string LearnMoreKey = "translation.common.learnMore";

    /// <summary>English fallback for <see cref="LearnMoreKey"/> (web second arg, verbatim).</summary>
    public const string LearnMoreFallback = "Learn more";

    /// <summary>ARIA role the revealed tooltip body declares (web shared <c>&lt;Tooltip role="tooltip"&gt;</c>).</summary>
    public const string TooltipRole = "tooltip";

    /// <summary>Segoe Fluent "Help" glyph — the native stand-in for the web Lucide <c>HelpCircle</c> trigger icon.</summary>
    public const string HelpGlyph = "\uE897";

    /// <summary>Segoe Fluent "OpenInNewWindow" glyph — the native stand-in for the web Lucide <c>ExternalLink</c> icon.</summary>
    public const string ExternalLinkGlyph = "\uE8A7";

    /// <summary>Trigger glyph font size for the extra-small tier (web <c>h-3 w-3</c> ≈ 12&#160;px).</summary>
    public const double ExtraSmallIconSize = 12;

    /// <summary>Trigger glyph font size for the small tier — the web default (web <c>h-3.5 w-3.5</c> ≈ 14&#160;px).</summary>
    public const double SmallIconSize = 14;

    /// <summary>Trigger glyph font size for the medium tier (web <c>h-4 w-4</c> ≈ 16&#160;px).</summary>
    public const double MediumIconSize = 16;

    /// <summary>Learn-more trailing glyph font size (web <c>ExternalLink h-3 w-3</c> ≈ 12&#160;px).</summary>
    public const double ExternalLinkIconSize = 12;

    /// <summary>The trigger glyph font size for a size tier (web <c>SIZE_CLASS</c> map).</summary>
    /// <param name="size">The web size tier.</param>
    public static double IconSize(HelpTooltipSize size) => size switch
    {
        HelpTooltipSize.ExtraSmall => ExtraSmallIconSize,
        HelpTooltipSize.Medium => MediumIconSize,
        _ => SmallIconSize,
    };
}

/// <summary>
/// The trigger glyph size tier — the native analogue of the web <c>size</c> prop (<c>'xs' | 'sm' | 'md'</c>,
/// <c>HelpTooltip.tsx</c> L28-29). <see cref="Small"/> is the web default.
/// </summary>
public enum HelpTooltipSize
{
    /// <summary>web <c>'xs'</c> — <c>h-3 w-3</c> (≈ 12&#160;px).</summary>
    ExtraSmall,

    /// <summary>web <c>'sm'</c> — <c>h-3.5 w-3.5</c> (≈ 14&#160;px); the default.</summary>
    Small,

    /// <summary>web <c>'md'</c> — <c>h-4 w-4</c> (≈ 16&#160;px).</summary>
    Medium,
}

/// <summary>
/// Where the revealed tooltip sits relative to the trigger — the native analogue of the web <c>placement</c>
/// prop (<c>'top' | 'bottom' | 'left' | 'right'</c>, <c>HelpTooltip.tsx</c> L25). <see cref="Top"/> is the web
/// default. The WinUI view maps this to a <c>PlacementMode</c> (hover tooltip) / <c>FlyoutPlacementMode</c>
/// (learn-more popup) at the platform boundary.
/// </summary>
public enum HelpTooltipPlacement
{
    /// <summary>web <c>'top'</c> — above the trigger; the default.</summary>
    Top,

    /// <summary>web <c>'bottom'</c> — below the trigger.</summary>
    Bottom,

    /// <summary>web <c>'left'</c> — to the left of the trigger.</summary>
    Left,

    /// <summary>web <c>'right'</c> — to the right of the trigger.</summary>
    Right,
}

/// <summary>
/// The optional "Learn more" affordance a help tooltip carries — the native analogue of the web <c>learnMore</c>
/// prop (<c>{ url: string; label?: string }</c>, <c>HelpTooltip.tsx</c> L27). It is pure data so the projection
/// stays unit-testable; the WinUI view materialises it as a <c>HyperlinkButton</c> whose <c>NavigateUri</c> opens
/// the external target (the web <c>target="_blank" rel="noopener noreferrer"</c> new tab). A missing
/// <see cref="Label"/> falls back to the localized <c>common.learnMore</c> string in the projection.
/// </summary>
public sealed record HelpTooltipLearnMore
{
    private HelpTooltipLearnMore(string url, string? label)
    {
        Url = url;
        Label = label;
    }

    /// <summary>The external target opened in a new tab (web <c>learnMore.url</c>).</summary>
    public string Url { get; }

    /// <summary>The optional override label (web <c>learnMore.label</c>); null falls back to the localized "Learn more".</summary>
    public string? Label { get; }

    /// <summary>Build a learn-more affordance.</summary>
    /// <param name="url">The external target (web <c>learnMore.url</c>); must be non-empty.</param>
    /// <param name="label">Optional override label (web <c>learnMore.label</c>); null uses the localized default.</param>
    public static HelpTooltipLearnMore Create(string url, string? label = null)
    {
        ArgumentException.ThrowIfNullOrEmpty(url);
        return new HelpTooltipLearnMore(url, label);
    }
}

/// <summary>
/// The render-time data model the <c>HelpTooltip</c> view binds to — the native analogue of the web
/// <c>HelpTooltipProps</c> (<c>HelpTooltip.tsx</c> L17-40). The web component is purely presentational: it resolves
/// its body from <see cref="Text"/> or the <see cref="I18nKey"/> / <see cref="DefaultValue"/> pair through
/// <c>useTranslation</c> and otherwise just reflects its props, so — exactly like React re-rendering the element
/// with already-resolved props — there is no fetch-driven loading / error / stale / offline branch to reproduce
/// here; the only branches are the resolved-vs-empty content (the web <c>if (!resolved) return null</c>), the three
/// size tiers, the four placements, the optional learn-more link, and the optional accessible-name override. The
/// trigger-icon override (web <c>children</c>) is a view concern (it is arbitrary content) surfaced through the
/// view's <c>TriggerContent</c> property rather than this data model. Pure data — no WinUI types — so the
/// projection is unit-tested without a UI host.
/// </summary>
public sealed record HelpTooltipModel
{
    private HelpTooltipModel(
        string? text,
        string? i18nKey,
        string? defaultValue,
        HelpTooltipSize size,
        HelpTooltipPlacement placement,
        HelpTooltipLearnMore? learnMore,
        string? ariaLabel)
    {
        Text = text;
        I18nKey = i18nKey;
        DefaultValue = defaultValue;
        Size = size;
        Placement = placement;
        LearnMore = learnMore;
        AriaLabel = ariaLabel;
    }

    /// <summary>Plain text body, used when no <see cref="I18nKey"/> is supplied (web <c>text</c>).</summary>
    public string? Text { get; }

    /// <summary>i18n key for the body; takes precedence over <see cref="Text"/> (web <c>i18nKey</c>).</summary>
    public string? I18nKey { get; }

    /// <summary>English fallback for <see cref="I18nKey"/> when the key is missing (web <c>defaultValue</c>).</summary>
    public string? DefaultValue { get; }

    /// <summary>The trigger glyph size tier (web <c>size</c>; defaults to <see cref="HelpTooltipSize.Small"/>).</summary>
    public HelpTooltipSize Size { get; }

    /// <summary>Where the tooltip sits relative to the trigger (web <c>placement</c>; defaults to <see cref="HelpTooltipPlacement.Top"/>).</summary>
    public HelpTooltipPlacement Placement { get; }

    /// <summary>Optional "Learn more" link (web <c>learnMore</c>); null renders the body only.</summary>
    public HelpTooltipLearnMore? LearnMore { get; }

    /// <summary>Optional accessible-name override for the trigger (web <c>ariaLabel</c>); null uses the localized "More info".</summary>
    public string? AriaLabel { get; }

    /// <summary>An empty tooltip with no resolvable content — the view collapses to nothing (web <c>return null</c>).</summary>
    public static HelpTooltipModel Empty { get; } = new(null, null, null, HelpTooltipSize.Small, HelpTooltipPlacement.Top, null, null);

    /// <summary>Build a help tooltip from plain text (web <c>text</c> path).</summary>
    /// <param name="text">The body text (web <c>text</c>).</param>
    /// <param name="size">The trigger size tier (web <c>size</c>).</param>
    /// <param name="placement">The tooltip placement (web <c>placement</c>).</param>
    /// <param name="learnMore">Optional learn-more link (web <c>learnMore</c>).</param>
    /// <param name="ariaLabel">Optional accessible-name override (web <c>ariaLabel</c>).</param>
    public static HelpTooltipModel FromText(
        string text,
        HelpTooltipSize size = HelpTooltipSize.Small,
        HelpTooltipPlacement placement = HelpTooltipPlacement.Top,
        HelpTooltipLearnMore? learnMore = null,
        string? ariaLabel = null)
    {
        ArgumentNullException.ThrowIfNull(text);
        return new HelpTooltipModel(text, null, null, size, placement, learnMore, ariaLabel);
    }

    /// <summary>Build a help tooltip from an i18n key + English fallback (web <c>i18nKey</c> / <c>defaultValue</c> path).</summary>
    /// <param name="i18nKey">The body i18n key (web <c>i18nKey</c>).</param>
    /// <param name="defaultValue">The English fallback when the key is missing (web <c>defaultValue</c>).</param>
    /// <param name="size">The trigger size tier (web <c>size</c>).</param>
    /// <param name="placement">The tooltip placement (web <c>placement</c>).</param>
    /// <param name="learnMore">Optional learn-more link (web <c>learnMore</c>).</param>
    /// <param name="ariaLabel">Optional accessible-name override (web <c>ariaLabel</c>).</param>
    public static HelpTooltipModel FromKey(
        string i18nKey,
        string defaultValue,
        HelpTooltipSize size = HelpTooltipSize.Small,
        HelpTooltipPlacement placement = HelpTooltipPlacement.Top,
        HelpTooltipLearnMore? learnMore = null,
        string? ariaLabel = null)
    {
        ArgumentException.ThrowIfNullOrEmpty(i18nKey);
        ArgumentNullException.ThrowIfNull(defaultValue);
        return new HelpTooltipModel(null, i18nKey, defaultValue, size, placement, learnMore, ariaLabel);
    }

    /// <summary>
    /// Build a help tooltip from the raw web prop shape, reproducing the web content precedence exactly
    /// (<c>HelpTooltip.tsx</c> L61-63): a non-empty <paramref name="i18nKey"/> resolves through translation (with
    /// <paramref name="defaultValue"/> as the fallback), otherwise the plain <paramref name="text"/> is used.
    /// </summary>
    /// <param name="text">The plain text body (web <c>text</c>); ignored when <paramref name="i18nKey"/> is set.</param>
    /// <param name="i18nKey">The body i18n key (web <c>i18nKey</c>); takes precedence over <paramref name="text"/>.</param>
    /// <param name="defaultValue">The English fallback for <paramref name="i18nKey"/> (web <c>defaultValue</c>).</param>
    /// <param name="size">The trigger size tier (web <c>size</c>).</param>
    /// <param name="placement">The tooltip placement (web <c>placement</c>).</param>
    /// <param name="learnMore">Optional learn-more link (web <c>learnMore</c>).</param>
    /// <param name="ariaLabel">Optional accessible-name override (web <c>ariaLabel</c>).</param>
    public static HelpTooltipModel Create(
        string? text = null,
        string? i18nKey = null,
        string? defaultValue = null,
        HelpTooltipSize size = HelpTooltipSize.Small,
        HelpTooltipPlacement placement = HelpTooltipPlacement.Top,
        HelpTooltipLearnMore? learnMore = null,
        string? ariaLabel = null) =>
        new(text, i18nKey, defaultValue, size, placement, learnMore, ariaLabel);
}

/// <summary>
/// The fully projected, render-ready view of a <see cref="HelpTooltipModel"/> — everything the web component
/// derives before returning JSX (<c>HelpTooltip.tsx</c> L59-113): the <see cref="ResolvedText"/> body (from the
/// i18n key or the plain text), whether there is <see cref="HasContent"/> (the web <c>if (!resolved) return
/// null</c> guard, surfaced as <see cref="RendersNothing"/>), the trigger <see cref="Glyph"/> + <see cref="IconSize"/>,
/// the composed <see cref="AccessibleLabel"/> Narrator reads (web <c>ariaLabel ?? t('help.tooltip.iconLabel')</c>),
/// the <see cref="Placement"/>, the optional learn-more affordance (<see cref="HasLearnMore"/> /
/// <see cref="LearnMoreUrl"/> / <see cref="LearnMoreLabel"/> / <see cref="ExternalLinkGlyph"/>), and the
/// <see cref="TooltipRole"/> the revealed body declares. Pure data so every value is asserted headlessly.
/// </summary>
/// <param name="ResolvedText">The resolved body text (web <c>resolved</c>); empty when there is nothing to show.</param>
/// <param name="HasContent">True when <see cref="ResolvedText"/> is non-empty (web <c>resolved</c> truthy).</param>
/// <param name="RendersNothing">True when there is no content — the view collapses to nothing (web <c>return null</c>).</param>
/// <param name="Glyph">The trigger glyph (web Lucide <c>HelpCircle</c> stand-in).</param>
/// <param name="IconSize">The trigger glyph font size for the size tier (web <c>SIZE_CLASS</c>).</param>
/// <param name="AccessibleLabel">The trigger's accessible name (web <c>ariaLabel ?? t('help.tooltip.iconLabel')</c>).</param>
/// <param name="Placement">Where the tooltip sits relative to the trigger (web <c>placement</c>).</param>
/// <param name="HasLearnMore">True when the learn-more link renders (web <c>learnMore</c> present).</param>
/// <param name="LearnMoreUrl">The learn-more external target (web <c>learnMore.url</c>); null when absent.</param>
/// <param name="LearnMoreLabel">The learn-more label (web <c>learnMore.label ?? t('common.learnMore')</c>); empty when absent.</param>
/// <param name="ExternalLinkGlyph">The trailing glyph on the learn-more link (web Lucide <c>ExternalLink</c> stand-in).</param>
/// <param name="TooltipRole">The ARIA role the revealed body declares (web <c>role="tooltip"</c>).</param>
public sealed record HelpTooltipDisplay(
    string ResolvedText,
    bool HasContent,
    bool RendersNothing,
    string Glyph,
    double IconSize,
    string AccessibleLabel,
    HelpTooltipPlacement Placement,
    bool HasLearnMore,
    string? LearnMoreUrl,
    string LearnMoreLabel,
    string ExternalLinkGlyph,
    string TooltipRole);

/// <summary>
/// Pure projection from a <see cref="HelpTooltipModel"/> to its <see cref="HelpTooltipDisplay"/> — the native port
/// of <c>web/src/components/ui/HelpTooltip.tsx</c>. Reproduces the web derivations exactly: the body is the
/// translated <c>i18nKey</c> (with <c>defaultValue</c> as the fallback) when a key is supplied, otherwise the plain
/// <c>text</c> (web L61-63); an empty body means the surface renders nothing (web L67 <c>if (!resolved) return
/// null</c>); the trigger glyph size comes from the size tier (web <c>SIZE_CLASS</c>, L42-46); the accessible name
/// is the caller's <c>ariaLabel</c> or the localized <c>help.tooltip.iconLabel</c> "More info" (web L70); and the
/// learn-more label is the caller's override or the localized <c>common.learnMore</c> "Learn more" (web L87). No
/// WinUI types — so the projection is unit-tested without a UI host, and the view binds to its result.
/// </summary>
public static class HelpTooltipProjection
{
    /// <summary>Project <paramref name="model"/> into a render-ready display, resolving strings through <paramref name="localizer"/>.</summary>
    /// <param name="model">The render-time data model (the web props).</param>
    /// <param name="localizer">The i18n facade (P1/S10) every string resolves through (web <c>useTranslation</c>).</param>
    /// <returns>The render-ready display model.</returns>
    public static HelpTooltipDisplay Project(HelpTooltipModel model, ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(model);
        ArgumentNullException.ThrowIfNull(localizer);

        // web L61-63: i18nKey ? t(i18nKey, { defaultValue }) : (text ?? '').
        string resolved = !string.IsNullOrEmpty(model.I18nKey)
            ? localizer.GetString(model.I18nKey, model.DefaultValue ?? string.Empty)
            : (model.Text ?? string.Empty);

        bool hasContent = !string.IsNullOrEmpty(resolved);

        // web L70: ariaLabel ?? t('help.tooltip.iconLabel', { defaultValue: 'More info' }).
        string accessibleLabel = !string.IsNullOrEmpty(model.AriaLabel)
            ? model.AriaLabel!
            : localizer.GetString(HelpTooltipRegistration.IconLabelKey, HelpTooltipRegistration.IconLabelFallback);

        HelpTooltipLearnMore? learnMore = model.LearnMore;
        bool hasLearnMore = learnMore is not null;

        // web L87: learnMore.label ?? t('common.learnMore', { defaultValue: 'Learn more' }).
        string learnMoreLabel = hasLearnMore
            ? (!string.IsNullOrEmpty(learnMore!.Label)
                ? learnMore.Label!
                : localizer.GetString(HelpTooltipRegistration.LearnMoreKey, HelpTooltipRegistration.LearnMoreFallback))
            : string.Empty;

        return new HelpTooltipDisplay(
            ResolvedText: resolved,
            HasContent: hasContent,
            RendersNothing: !hasContent,
            Glyph: HelpTooltipRegistration.HelpGlyph,
            IconSize: HelpTooltipRegistration.IconSize(model.Size),
            AccessibleLabel: accessibleLabel,
            Placement: model.Placement,
            HasLearnMore: hasLearnMore,
            LearnMoreUrl: learnMore?.Url,
            LearnMoreLabel: learnMoreLabel,
            ExternalLinkGlyph: HelpTooltipRegistration.ExternalLinkGlyph,
            TooltipRole: HelpTooltipRegistration.TooltipRole);
    }
}

/// <summary>
/// PII-safe diagnostics for the <c>HelpTooltip</c> surface (P1/S11 diagnostics contract). Help bodies can carry
/// user-facing copy, so the collector records only the operational <c>view.opened</c> event with the surface slug
/// — never the resolved text, accessible label or learn-more URL — so a diagnostics line can never leak fleet
/// state. Thread-safe; mirrors the peer surfaces' diagnostics collectors.
/// </summary>
public sealed class HelpTooltipDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    /// <param name="sink">An optional sink the operational event line is written to.</param>
    public HelpTooltipDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=HelpTooltip</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={HelpTooltipRegistration.Slug}");
    }
}

File: apps/android/app/src/main/java/com/teslasync/shared-surfaces/HelpTooltip/HelpTooltipModel.kt
// Pure, framework-free model + projection + diagnostics for the HelpTooltip shared surface — the native
// analogue of web/src/components/ui/HelpTooltip.tsx. No Compose, no Android framework, no HTTP: every
// declaration here is exercised off-device in the :android:testReleaseUnitTest gate, keeping the composable a
// thin render layer (ADR-002).
//
// The web source is a PRESENTATIONAL leaf, not a data-fetching view: its only bound hook is `useTranslation`
// and it fetches nothing. It renders a compact "?" trigger that, on hover / focus / tap, reveals an
// explanatory tooltip carrying the resolved body copy and an optional "Learn more" link that opens in a new
// tab. Because the surface has no async cache-then-network feed, there is no loading / empty / error / stale /
// offline lifecycle to invent; modelling those would fabricate behaviour the web spec does not have (the same
// rationale the accepted HelpSegment / CopyLinkButton / GuardedLink / VisuallyHidden ports document, covenant
// #2 / #9). The surface's REAL states are reproduced instead and modelled here as pure, testable declarations:
//   • content ABSENT — `i18nKey ? t(i18nKey, {defaultValue}) : text` resolves to empty, so the web returns
//     `null` and renders nothing ([hasHelpContent] false); the native surface renders nothing too;
//   • content PRESENT, no link — just the resolved body;
//   • content PRESENT + "Learn more" — the body plus the external link;
//   • the trigger [HelpTooltipSize] (xs / sm / md, web `SIZE_CLASS`) and [HelpTooltipPlacement]
//     (top / bottom / left / right, web `placement`) variants;
//   • the link-open [LinkOutcome] — the platform opened the URL or rejected it (the native analogue of a web
//     new-tab navigation succeeding or being blocked).
// Every rendered string resolves through the i18n catalog (P1/S10) at the render boundary via the web
// `t(key, default)` contract, so no un-internationalized literal lives in native code — the [ICON_LABEL_KEY] /
// [LEARN_MORE_KEY] keys + their fallbacks below are the web source's own keys and i18next default values.
//
// `InvalidPackageDeclaration` is suppressed because the mandated surface directory
// (com/teslasync/shared-surfaces/HelpTooltip — the P3 prompt's allowed-files path) cannot form a valid Kotlin
// package (a hyphen segment and a PascalCase leaf are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling surfaces do. `MatchingDeclarationName` is
// suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.sharedsurfaces.helptooltip

import io.teslasync.shared.core.diagnostics.Logger

/**
 * Canonical registry metadata for the HelpTooltip surface. The diagnostics [SLUG] is emitted with the
 * one-shot `view.opened` event (P1/S11) and is the surface slug the prompt mandates (`HelpTooltip`); [ID] is
 * the stable `viewModel` key the composable binds its state holder with, and the test tags name the trigger
 * and "Learn more" nodes the UI test drives.
 */
object HelpTooltipRegistration {
    /** Stable surface id, also the `viewModel` key the composable binds its holder with. */
    const val ID: String = "help-tooltip"

    /** Diagnostics surface slug emitted with the `view.opened` / `helpTooltip.learnMore` events (P1/S11). */
    const val SLUG: String = "HelpTooltip"

    /** Test tag for the "?" trigger node, present in every render state that shows content. */
    const val TRIGGER_TEST_TAG: String = "help-tooltip-trigger"

    /** Test tag for the optional "Learn more" affordance inside the tooltip. */
    const val LEARN_MORE_TEST_TAG: String = "help-tooltip-learn-more"
}

/** The web `t('help.tooltip.iconLabel', 'More info')` key — the trigger's default accessible name. */
const val ICON_LABEL_KEY: String = "help.tooltip.iconLabel"

/** The i18next default for [ICON_LABEL_KEY] (the web source's inline fallback). */
const val ICON_LABEL_FALLBACK: String = "More info"

/** The web `t('common.learnMore', 'Learn more')` key — the default "Learn more" link label. */
const val LEARN_MORE_KEY: String = "common.learnMore"

/** The i18next default for [LEARN_MORE_KEY] (the web source's inline fallback). */
const val LEARN_MORE_FALLBACK: String = "Learn more"

/**
 * The optional "Learn more" affordance shown below the tooltip body — the native analogue of the web
 * `learnMore: { url: string; label?: string }` prop. [url] is opened externally through the [LinkOpener] seam
 * (web `<a target="_blank">`); [label] overrides the default `t('common.learnMore')` link text when supplied.
 * A pure data holder so it can be constructed and asserted off-device.
 *
 * @property url the external link opened on tap (web `learnMore.url`).
 * @property label the optional link label; falls back to [LEARN_MORE_FALLBACK] / the catalog when `null`.
 */
data class HelpTooltipLearnMore(
    val url: String,
    val label: String? = null,
)

/**
 * The trigger icon size — the native tag for the web `size` prop and its `SIZE_CLASS` map. Each carries the
 * web pixel dimension as [iconDp] (web xs = `h-3 w-3` = 12, sm = `h-3.5` = 14, md = `h-4` = 16) so the
 * composable maps it onto the shared `IconSize` and the dimension is unit-tested off-device.
 *
 * @property iconDp the trigger glyph edge length in dp, mirroring the web Tailwind size class.
 */
enum class HelpTooltipSize(
    val iconDp: Int,
) {
    /** Web `size="xs"` → `h-3 w-3` (12 dp). */
    Xs(12),

    /** Web `size="sm"` → `h-3.5 w-3.5` (14 dp) — the web default. */
    Sm(14),

    /** Web `size="md"` → `h-4 w-4` (16 dp). */
    Md(16),
}

/**
 * Where the tooltip is placed relative to the trigger — the native tag for the web `placement` prop (passed
 * to the shared `Tooltip` as `side`). All four sides are honoured by the surface's position provider via the
 * pure [helpTooltipPopupOffset] geometry below.
 */
enum class HelpTooltipPlacement {
    /** Above the trigger — the web default. */
    Top,

    /** Below the trigger. */
    Bottom,

    /** To the (physical) left of the trigger; swaps with [Right] under RTL. */
    Left,

    /** To the (physical) right of the trigger; swaps with [Left] under RTL. */
    Right,
}

/**
 * Resolves the requested [placement] to the physical side actually used, swapping [HelpTooltipPlacement.Left]
 * and [HelpTooltipPlacement.Right] under [isRtl] so a "left"/"right" tooltip stays on the reading-order side —
 * the standard Compose layout-direction convention. [HelpTooltipPlacement.Top] / [HelpTooltipPlacement.Bottom]
 * are unaffected. Pure so the mirroring is unit-tested without a UI host.
 */
fun resolvePhysicalPlacement(
    placement: HelpTooltipPlacement,
    isRtl: Boolean,
): HelpTooltipPlacement =
    when (placement) {
        HelpTooltipPlacement.Left -> if (isRtl) HelpTooltipPlacement.Right else HelpTooltipPlacement.Left
        HelpTooltipPlacement.Right -> if (isRtl) HelpTooltipPlacement.Left else HelpTooltipPlacement.Right
        else -> placement
    }

/**
 * The web content-resolution precedence — `const resolved = i18nKey ? t(i18nKey, {defaultValue}) : text`.
 * When an [i18nKey] is supplied the body is the [catalogValue] the composable resolved from the P1/S10 catalog
 * (or `null` when the key is catalog-absent), falling back to [defaultValue]; otherwise it is the literal
 * [text]. A `null` input collapses to the empty string, exactly as the web `?? ''` / `text ?? ''` do. Pure so
 * the precedence is unit-tested off-device, free of the catalog lookup.
 */
fun resolveHelpBody(
    text: String?,
    i18nKey: String?,
    defaultValue: String?,
    catalogValue: String?,
): String =
    if (i18nKey != null) {
        catalogValue ?: defaultValue ?: ""
    } else {
        text ?: ""
    }

/**
 * Whether there is body content to render — the native mirror of the web `if (!resolved) return null`. Uses
 * `isNotEmpty` (not `isNotBlank`) to match JavaScript truthiness, where an empty string is falsy but a
 * whitespace-only string is truthy. When this is false the surface renders nothing.
 */
fun hasHelpContent(resolved: String): Boolean = resolved.isNotEmpty()

/**
 * The visible "Learn more" link label — the web `learnMore.label ?? t('common.learnMore', 'Learn more')`. A
 * supplied [custom] label wins (matching the `??` null-coalesce, so only a missing label falls back); otherwise
 * the catalog-resolved [fallback] is used. Pure so the choice is unit-tested off-device.
 */
fun resolveLearnMoreLabel(
    custom: String?,
    fallback: String,
): String = custom ?: fallback

/**
 * A framework-free 2-D integer offset (top-left of the tooltip popup). The composable maps it onto a Compose
 * `IntOffset`; kept Compose-free so the placement geometry is unit-tested off-device.
 */
data class HelpTooltipOffset(
    val x: Int,
    val y: Int,
)

/**
 * Computes the tooltip popup's top-left position for [placement] relative to the trigger — the pure geometry
 * behind the surface's custom position provider, so all four web placements (not just Material's default
 * above/below) are honoured and unit-tested off-device.
 *
 * The popup is centred on the trigger's cross axis and offset by [gap] px on the main axis, then clamped into
 * the `[0, windowSize - popupSize]` window box so it can never spill off-screen. [isRtl] mirrors left/right via
 * [resolvePhysicalPlacement]. All inputs are integer pixels (the composable supplies the anchor bounds, popup
 * content size, and window size from the Material `TooltipBox` position-provider callback).
 */
@Suppress("LongParameterList")
fun helpTooltipPopupOffset(
    placement: HelpTooltipPlacement,
    anchorLeft: Int,
    anchorTop: Int,
    anchorWidth: Int,
    anchorHeight: Int,
    popupWidth: Int,
    popupHeight: Int,
    windowWidth: Int,
    windowHeight: Int,
    gap: Int,
    isRtl: Boolean,
): HelpTooltipOffset {
    val side = resolvePhysicalPlacement(placement, isRtl)
    val anchorRight = anchorLeft + anchorWidth
    val anchorBottom = anchorTop + anchorHeight
    val centerX = anchorLeft + anchorWidth / 2
    val centerY = anchorTop + anchorHeight / 2

    val rawX: Int
    val rawY: Int
    when (side) {
        HelpTooltipPlacement.Top -> {
            rawX = centerX - popupWidth / 2
            rawY = anchorTop - popupHeight - gap
        }
        HelpTooltipPlacement.Bottom -> {
            rawX = centerX - popupWidth / 2
            rawY = anchorBottom + gap
        }
        HelpTooltipPlacement.Left -> {
            rawX = anchorLeft - popupWidth - gap
            rawY = centerY - popupHeight / 2
        }
        HelpTooltipPlacement.Right -> {
            rawX = anchorRight + gap
            rawY = centerY - popupHeight / 2
        }
    }

    val maxX = (windowWidth - popupWidth).coerceAtLeast(0)
    val maxY = (windowHeight - popupHeight).coerceAtLeast(0)
    return HelpTooltipOffset(rawX.coerceIn(0, maxX), rawY.coerceIn(0, maxY))
}

/**
 * The resolved outcome of opening the "Learn more" link — emitted (PII-free) as a diagnostics field so a link
 * open can be observed without ever recording the URL (which can carry deep-link state). [Opened] is the
 * platform accepting the new-tab navigation; [Failed] is it rejecting it (no browser / activity).
 */
enum class LinkOutcome(
    val wireName: String,
) {
    /** The platform opened the link (web: the new tab launched). */
    Opened("opened"),

    /** The platform rejected the open (web: the navigation was blocked / no handler). */
    Failed("failed"),
}

/** Maps a link-open result onto its diagnostics [LinkOutcome]. */
fun linkOutcomeFor(succeeded: Boolean): LinkOutcome = if (succeeded) LinkOutcome.Opened else LinkOutcome.Failed

/** The stable, dot-namespaced diagnostics event emitted once when the surface is first composed (P1/S11). */
const val EVENT_VIEW_OPENED: String = "view.opened"

/** The diagnostics event emitted (PII-free) whenever the "Learn more" link is opened. */
const val EVENT_LEARN_MORE: String = "helpTooltip.learnMore"

/** The structured-field key carrying the surface slug on every diagnostic. */
const val FIELD_SURFACE: String = "surface"

/** The structured-field key carrying the link-open outcome (never the opened URL). */
const val FIELD_OUTCOME: String = "outcome"

/**
 * PII-safe diagnostics for the HelpTooltip surface (P1/S11). Every record carries only the surface
 * [HelpTooltipRegistration.SLUG] and, for a link open, the coarse [LinkOutcome] — never the body copy, the
 * label, or the opened URL, so a diagnostics line can never leak help copy or where a user navigated. Kept
 * free of Compose so it is unit-tested with a recording [Logger].
 */
object HelpTooltipDiagnostics {
    /** Emits the one `view.opened` record (slug only) — the ViewModel calls it once per surface open. */
    fun recordViewOpened(logger: Logger) {
        logger.info(EVENT_VIEW_OPENED, mapOf(FIELD_SURFACE to HelpTooltipRegistration.SLUG))
    }

    /** Emits the `helpTooltip.learnMore` record carrying the surface slug and the coarse [outcome] only. */
    fun recordLearnMore(
        logger: Logger,
        outcome: LinkOutcome,
    ) {
        logger.info(
            EVENT_LEARN_MORE,
            mapOf(
                FIELD_SURFACE to HelpTooltipRegistration.SLUG,
                FIELD_OUTCOME to outcome.wireName,
            ),
        )
    }
}
